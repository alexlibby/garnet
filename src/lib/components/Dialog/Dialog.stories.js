import Dialog from "./Dialog.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Notification Components/Dialog",
  component: Dialog,
  argTypes: {
    showDialog: { action: "boolean" },
    showBackground: { action: "boolean" },
    modalTitle: { action: "string" },
    modalText: { action: "string" },
  },
};

export const Default = {
  component: Dialog,
  args: {
    showDialog: true,
    modalTitle: "Modal title",
    modalText: "Click on the X to close me",
  },
  parameters: {
    docs: {
      story: {
        height: "300px",
        width: "300px",
      },
    },
  },
};

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const ClosedOnOpen = {
  component: Dialog,
  args: {
    showDialog: false,
    modalTitle: "Modal title",
    modalText: "This modal is closed on launch",
  },
  parameters: {
    docs: {
      story: {
        height: "300px",
        width: "300px",
      },
    },
  },
};

ClosedOnOpen.parameters = {
  badges: [BADGES.ALPHA],
};

export const NoBackground = {
  component: Dialog,
  args: {
    showDialog: false,
    showBackground: false,
    modalTitle: "Modal title",
    modalText: "This modal doesn't show a background",
  },
  parameters: {
    docs: {
      story: {
        height: "300px",
        width: "300px",
      },
    },
  },
};

NoBackground.parameters = {
  badges: [BADGES.ALPHA],
};
