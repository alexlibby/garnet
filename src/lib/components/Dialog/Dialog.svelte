<svelte:options customElement="garnet-dialog" />

<script>
  import CloseIcon from "./Close.svelte";
  export let showDialog = true;
  export let showBackground = false;
  export let modalTitle = "Modal title";
  export let modalText = "Click on the X to close me";
</script>

<button data-testid = "buttonDialog" on:click={() => (showDialog = !showDialog)}>Show dialog</button>

{#if showDialog}
  <div class="garnet">
    {#if showBackground}<div id="background" />{/if}
    <div data-testid="modal" id="modal">
      <div class="header">
        <h3>{modalTitle}</h3>
        <button
          type="button"
          class="close"
          data-testid="close-button"
          title="close button"
          on:click={() => (showDialog = false)}
        >
          <CloseIcon />
        </button>
      </div>
      <p>{modalText}</p>
    </div>
  </div>
{/if}

<style>
  .garnet {
    font-family: Arial, Helvetica, sans-serif;
  }

  @keyframes fadein {
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  }

  #background {
    position: fixed;
    z-index: 1;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background-color: rgba(0, 0, 0, 0.7);
    backdrop-filter: blur(5px);
    animation: fadein 1s;
  }

  #modal {
    position: fixed;
    z-index: 2;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background: #fff;
    padding: 10px;
    width: 400px;
    height: 250px;
    border: 1px solid #c4c4c4;
    box-shadow: 2.5px 5.0px 5.0px hsl(0deg 0% 0% / 0.42);
  }

  .header {
    display: flex;
    justify-content: space-between;
    border-bottom: 1px solid #c4c4c4;
  }

  button {
    background-color: #733635;
    border: none;
    color: white;
    padding: 15px 32px;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 16px;  
  }

 button.close {
    background-color: #ffffff;
    padding: 10px 0 10px 0;    
  }

  button.close:hover {
    color: #d19c9b;
  }
</style>
