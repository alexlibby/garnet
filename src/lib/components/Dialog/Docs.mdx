import { Canvas, Meta, Story } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as DialogStories from "./Dialog.stories";

<Meta of={DialogStories} />

# Dialog

Dialog is the primary component. It has three possible states.

- [Default](#default)
- [Closed On Open](#closedonopen)
- [NoBackground](#nobackground)

## Default

This is the default state for the Dialog component:

<Canvas of={DialogStories.Default} />

## ClosedOnOpen

This state shows the Dialog closed, when launched:

<Canvas of={DialogStories.ClosedOnOpen} />

## NoBackground

This state shows the Dialog displayed without a background:

<Canvas of={DialogStories.NoBackground} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={DialogStories} />
