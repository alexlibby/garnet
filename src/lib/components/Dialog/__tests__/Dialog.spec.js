import { describe, test, expect, vi } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Dialog from "../../Dialog/Dialog.svelte";

const DialogHTML = `<div class="garnet s-ZH7mGohnGwmj"><div id="background" class="s-ZH7mGohnGwmj"></div> <div data-testid="modal" id="modal" class="s-ZH7mGohnGwmj"><div class="header s-ZH7mGohnGwmj"><h3 class="s-ZH7mGohnGwmj">Modal title</h3> <button type="button" class="close s-ZH7mGohnGwmj" data-testid="close-button" title="close button"><button class="close-button s-Pc0SWcMC6Bof" aria-label="Close"><svg class="close s-Pc0SWcMC6Bof" viewBox="0 0 24 24" aria-hidden="true" style="color: inherit;"><path fill="currentColor" d="M.439 21.44a1.5 1.5 0 0 0 2.122 2.121l9.262-9.261a.25.25 0 0 1 .354 0l9.262 9.263a1.5 1.5 0 1 0 2.122-2.121L14.3 12.177a.25.25 0 0 1 0-.354l9.263-9.262A1.5 1.5 0 0 0 21.439.44L12.177 9.7a.25.25 0 0 1-.354 0L2.561.44A1.5 1.5 0 0 0 .439 2.561L9.7 11.823a.25.25 0 0 1 0 .354Z" class="s-Pc0SWcMC6Bof"></path></svg></button><!--<Close>--></button></div> <p class="s-ZH7mGohnGwmj">Click on the X to close me</p></div></div>`;
expect.extend(AxeMatchers);

describe("Tests for Dialog", () => {
  const props = {
    showDialog: "false",
  };

  it("should render properly", async () => {
    const result = render(Dialog, { props });
    expect(() => result).not.toThrow();
  });

  it("should show the dialog when clicking the button", async () => {
    render(Dialog, { props });

    const button = screen.getByText("Show dialog");
    fireEvent.click(button);

    const modal = screen.getByTestId("modal");
    expect(modal).toBeInTheDocument();
  });

  it("should close the dialog when clicking the close button", async () => {
    render(Dialog, { props });

    // opens dialog
    const button = screen.getByText("Show dialog");
    fireEvent.click(button);

    // closes dialog
    const modal = screen.getByTestId("modal");
    const close = screen.getByTestId("close-button");
    await fireEvent.click(close);
    expect(modal).not.toBeInTheDocument();
  });

  it("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Dialog({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => DialogHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
