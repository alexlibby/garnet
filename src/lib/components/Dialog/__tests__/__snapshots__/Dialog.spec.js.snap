// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Dialog > get a snapshot of component 1`] = `"<button data-testid="buttonDialog" class="svelte-1u0li5y">Show dialog</button> <div class="garnet-dialog svelte-1u0li5y"> <div data-testid="modal" id="modal" class="svelte-1u0li5y"><div class="header svelte-1u0li5y"><h3>Modal title</h3> <button type="button" class="close svelte-1u0li5y" data-testid="close-button" title="close button"><button class="close-button svelte-myhoof" aria-label="Close"><svg class="close" viewBox="0 0 24 24" aria-hidden="true"><path fill="currentColor" d="M.439 21.44a1.5 1.5 0 0 0 2.122 2.121l9.262-9.261a.25.25 0 0 1 .354 0l9.262 9.263a1.5 1.5 0 1 0 2.122-2.121L14.3 12.177a.25.25 0 0 1 0-.354l9.263-9.262A1.5 1.5 0 0 0 21.439.44L12.177 9.7a.25.25 0 0 1-.354 0L2.561.44A1.5 1.5 0 0 0 .439 2.561L9.7 11.823a.25.25 0 0 1 0 .354Z"></path></svg></button></button></div> <p>Click on the X to close me</p></div></div>"`;
